src/WpfMath/TexFormulaParser.cs

```diff
@@ -436,11 +585,209 @@ private TexFormula ReadScript(TexFormula formula, SourceSpan value, ref int posi
+                {
+                    List<Atom> rowcellatoms = new List<Atom>();
+                    for (int j = 0; j < rowitem.Count; j++)
+                    {
+                        var cellitem = rowdata[i][j];
+                        if (cellitem.ToString().Trim().Length > 0 || cellitem.Length > 0)
+                        {
+                            var cellsource = matrixsource.Segment(matrixsrcstart, cellitem.Length);// new SourceSpan(cellitem, matrixsrcstart, cellitem.Length);
+                            var cellformula = Parse(cellsource, formula.TextStyle);
+                            if (allowEmptyCells)
+                            {
+                                if (cellformula.RootAtom == null)
+                                {
+                                    cellsource = new SourceSpan(" ", 0, 1);
+                                    rowcellatoms.Add(new NullAtom(cellsource));
+                                }
+                                else
+                                {
+                                    rowcellatoms.Add(cellformula.RootAtom);
+                                }
+                            }
+                            else
+                            {
+                                rowcellatoms.Add(cellformula.RootAtom);
+                            }
+
+                        }
+                        else
+                        {
+                            //Compensate by adding an invincible whitespace
+                            //(This allows for empty cells instead of relying on the user to request a phantom atom or space atom)
+                            var cellsource = new SourceSpan(" ", 0, 1);
+                            rowcellatoms.Add(new NullAtom(cellsource));
+                        }
+                        if (j < (rowitem.Count - 1))
+                        {
+                            matrixsrcstart += (cellitem.Length + 1);
+                        }
+                        else
+                        {
+                            matrixsrcstart += (cellitem.Length + rowindent + 1);
+                        }
+                    }
+
+                    matrixcells.Add(rowcellatoms);
+                    columnscount = rowcellatoms.Count;
+                }
+
+            }
+
+            int colsvalid = 0;
+            foreach (var item in matrixcells)
+            {
+                if (item.Count == columnscount)
+                {
+                    colsvalid++;
+                }
+            }
+
+            if (parseasArrays)
+            {
+                return matrixcells;
+            }
+            else
+            {
+                if (colsvalid == matrixcells.Count)
+                {
+                    return matrixcells;
+                }
+                else
+                {
+                    throw new TexParseException("The column numbers are not equal.");
+                }
+            }
+        }
```

Comment:
This method is very big and should be reduced into smaller and more testable pieces. Also it seems that it's missing `&`… somewhere. Will need additional debugging.

Reply:
Looks like I've fixed the issue with `&`.

Reply:
To fix this method, I'll need to partially implement #157 first.